src/axom/inlet/examples/user_defined_type.cpp

```diff
@@ -0,0 +1,139 @@
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "axom/inlet.hpp"
+
+#include <iostream>
+#include "CLI11/CLI11.hpp"
+#include "axom/slic/core/UnitTestLogger.hpp"
+
+using axom::inlet::Inlet;
+using axom::inlet::LuaReader;
+using axom::inlet::SchemaCreator;
+using axom::sidre::DataStore;
+
+struct Mesh
+{
+  std::string filename;
+  int serial_ref_iter;
+  int par_ref_iter;
+
+  // Each class should define a static method that adds the fields it
+  // will grab from inlet
+  static void defineSchema(SchemaCreator& schema)
+  {
+    schema.addString("filename", "Path to mesh file");
+    schema.addInt("serial", "Number of serial refinement iterations");
+    schema.addInt("parallel", "Number of parallel refinement iterations");
+  }
+};
+
+// Additionally, each class should specialize this struct as follows
+// in the global namespace so that Inlet can access it
+template <>
+struct FromInlet<Mesh>
+{
+  Mesh operator()(axom::inlet::Table& base)
+  {
+    return {base["filename"], base["serial"], base["parallel"]};
+  }
+};
+
+struct LinearSolver
+{
+  double rel_tol;
+  double abs_tol;
+  int print_level;
+  int max_iter;
+  double dt;
+  int steps;
+  static void defineSchema(SchemaCreator& schema)
+  {
+    schema.addDouble("rel_tol", "Relative convergence criterion");
+    schema.addDouble("abs_tol", "Relative convergence criterion");
+    schema.addInt("print_level", "Logging level for iterative solver");
+    schema.addInt("max_iter", "Maximum iteration count");
+    schema.addDouble("dt", "Time step");
+    schema.addInt("steps", "Number of simulation iterations/frames");
+  }
+};
+
+template <>
+struct FromInlet<LinearSolver>
+{
+  LinearSolver operator()(axom::inlet::Table& base)
+  {
+    LinearSolver lin_solve;
+    lin_solve.rel_tol = base["rel_tol"];
+    lin_solve.abs_tol = base["abs_tol"];
+    lin_solve.print_level = base["print_level"];
+    lin_solve.max_iter = base["max_iter"];
+    lin_solve.dt = base["dt"];
+    lin_solve.steps = base["steps"];
+    return lin_solve;
+  }
+};
+
+struct ThermalSolver
+{
+  Mesh mesh;
+  LinearSolver solver;
+  // defineSchema is intended to be used recursively
+  // Tables are created for subobjects and passed to
+  // subobject defineSchema implementations
+  static void defineSchema(SchemaCreator& schema)
+  {
+    auto mesh_table = schema.addTable("mesh", "Information about the mesh");
+    Mesh::defineSchema(*mesh_table);
+    auto solver_table =
+      schema.addTable("solver",
+                      "Information about the iterative solver used for Ku = f");
+    LinearSolver::defineSchema(*solver_table);
+  }
+};
+
+template <>
+struct FromInlet<ThermalSolver>
+{
+  // This is also implicitly recursive - will call the from_inlet
+  // functions defined for the subobjects
+  ThermalSolver operator()(axom::inlet::Table& base)
+  {
+    return {base["mesh"].get<Mesh>(), base["solver"].get<LinearSolver>()};
```

Comment:
This is really nice!

Reply:
Would it be possible to add an array case? Or is that not yet ready?

Reply:
This is in-progress on https://github.com/LLNL/axom/tree/feature/essman/inlet_record_arrays but I wanted to keep that functionality as a separate PR since this one is already rather large, in my opinion.